packages/flutter/lib/src/cupertino/colors.dart

```diff
@@ -543,41 +541,106 @@ class CupertinoColors {
+/// mode, and changes automatically to black in dark mode.
+///
+/// ```dart
+/// Builder(builder: (BuildContext context) {
+///   return CupertinoButton(
+///     child: child,
+///     // CupertinoDynamicColor works out of box in a CupertinoButton.
+///     color: CupertinoDynamicColor.withBrightness(
+///       color: CupertinoColors.white,
+///       darkColor: CupertinoColors.black,
+///     ),
+///     onPressed: () => null,
+///   );
+/// })
+/// ```
+/// {@end-tool}
+///
+/// ### Using a [CupertinoDynamicColor] from a [CupertinoTheme]
+///
+/// When referring to a [CupertinoTheme] color, generally the color will already
+/// have adapted to the ambient [BuildContext], because [CupertinoTheme.of]
+/// implicitly resolves all the colors used in the retrieved [CupertinoThemeData],
+/// before returning it.
+///
+/// {@tool sample}
+/// The following snippet creates a [Container] with the `primaryColor` of the
+/// current theme. If `primaryColor` is a [CupertinoDynamicColor], the container
+/// will be adaptive, thanks to [CupertinoTheme.of]: it will switch to `primaryColor`'s
+/// dark variant once dark mode is turned on, and turns to primaryColor`'s high
+/// contrast variant when [MediaQueryData.highContrast] is requested in the ambient
+/// [MediaQuery], etc.
+///
+/// ```dart
+/// Builder(builder:(BuildContext context) {
+///   // Container is not a Cupertino widget, but CupertinoTheme.of implicitly resolves
+///   // colors used in the retrieved CupertinoThemeData.
+///   return Container(
+///     color: CupertinoTheme.of(context).primaryColor,
+///   );
+/// })
+/// ```
+/// {@end-tool}
+///
+/// ### Manually Resolving a [CupertinoColor]
+///
+/// When used outside of a Cupertino component, or wrapped in an object opaque to
+/// the receiving Cupertino component, you may have to manually resolve the color
+/// using [CupertinoDynamicColor.resolve] before painting. For example, to use a
+/// custom [Border] in a [CupertinoNavigationBar], the colors used in the [Border]
+/// have to be resolved manually before being passed to [CupertinoNavigationBar]'s
+/// constructor.
+///
+/// {@tool sample}
+///
+/// The following snippet demostrates two cases where you have to manually resolve
+/// a [CupertinoDynamicColor]: [CupertinoColors.systemBlue].
+///
+/// ```dart
+/// Builder(builder: (BuildContext context) {
+///   final Color resolvedColor = CupertinoDynamicColor.resolve(CupertinoColors.systemBlue, context);
+///   return CupertinoNavigationBar(
+///     // CupertinoNavigationBar does not know how to resolve colors used in
+///     // a Border class.
+///     border: Border(bottom: BorderSide(color: resolvedColor)),
+///   );
+/// })
+/// ```
+///
+/// ```dart
+/// Builder(builder: (BuildContext context) {
+///   final Color resolvedColor = CupertinoDynamicColor.resolve(CupertinoColors.systemBlue, context);
+///   // Container is not a Cupertino widget.
+///   return Container(color: resolvedColor);
+/// })
+/// ```
+/// {@end-tool}
```

Comment:
Yes exactly. As of now all Cupertino widgets (with 1 or 2 exceptions I think, some PRs are blocked or still waiting to be merged) resolve the colors given to them in their `build` or `createRenderObject`/`updateRenderObject` methods. But if a Cupertino widget takes a `Border` things get complicated because it can be a subclass of `Border`. I feel this is going to be a big gotcha but not sure how I can better document it.

Reply:
What happens if you do pass an unresolved CupertinoDynamicColor to Border, does it use the light mode color as default?  Would it be possible to log a warning in that case, or does it happen too commonly in other situations?

Now that I understand the problem the examples above are pretty clear, though.  Maybe bring this up to the team and see if anyone has any ideas if you're still worried about it.

Reply:
Good idea! It sounds wrong to me to use a `CupertinoDynamicColor` that has never been resolved directly. Maybe putting some asserts in `CupertinoDynamicColor.value` override to ensure no one does something like `Paint()..color = CupertinoColors.systemBlue` will be a viable approach. I'll try this out in a different PR to see if that breaks anything.